bayesflow/experimental/graphical_simulator/graphical_simulator.py

```diff
@@ -0,0 +1,199 @@
+import inspect
+import itertools
+from collections.abc import Callable
+from typing import Any, Optional
+
+import networkx as nx
+import numpy as np
+
+from bayesflow.simulators import Simulator
+from bayesflow.types import Shape
+from bayesflow.utils.decorators import allow_batch_size
+
+
+class GraphicalSimulator(Simulator):
+    """
+    A graph-based simulator that generates samples by traversing a DAG
+    and calling user-defined sampling functions at each node.
+
+    Parameters
+    ----------
+    meta_fn : Optional[Callable[[], dict[str, Any]]]
+        A callable that returns a dictionary of meta data.
+        This meta data can be used to dynamically vary the number of sampling repetitions (`reps`)
+        for nodes added via `add_node`.
+    """
+
+    def __init__(self, meta_fn: Optional[Callable[[], dict[str, Any]]] = None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.graph = nx.DiGraph()
+        self.meta_fn = meta_fn
+
+    def add_node(self, node: str, sample_fn: Callable[..., dict[str, Any]], reps: int | str = 1):
+        self.graph.add_node(node, sample_fn=sample_fn, reps=reps)
+
+    def add_edge(self, from_node: str, to_node: str):
+        self.graph.add_edge(from_node, to_node)
+
+    @allow_batch_size
+    def sample(self, batch_shape: Shape | int, **kwargs) -> dict[str, np.ndarray]:
+        """
+        Generates samples by topologically traversing the DAG.
+        For each node, the sampling function is called based on parent values.
+
+        Parameters
+        ----------
+        batch_shape : Shape
+            The shape of the batch to sample. Typically, a tuple indicating the number of samples,
+            but an int can also be passed.
+        **kwargs
+            Unused
+        """
+        _ = kwargs  # Simulator class requires **kwargs, which are unused here
+        meta_dict = self.meta_fn() if self.meta_fn else {}
+        samples_by_node = {}
+
+        # Initialize samples container for each node
+        for node in self.graph.nodes:
+            samples_by_node[node] = np.empty(batch_shape, dtype="object")
+
+        for batch_idx in np.ndindex(batch_shape):
+            for node in nx.topological_sort(self.graph):
+                node_samples = []
+
+                parent_nodes = list(self.graph.predecessors(node))
+                sampling_fn = self.graph.nodes[node]["sample_fn"]
+                reps_field = self.graph.nodes[node]["reps"]
+                reps = reps_field if isinstance(reps_field, int) else meta_dict[reps_field]
+
+                if not parent_nodes:
+                    # root node: generate independent samples
+                    node_samples = [
+                        {"__batch_idx": batch_idx, f"__{node}_idx": i} | self._call_sampling_fn(sampling_fn, {})
```

Comment:
given that the UI now say "sample_fn" shall all the internals also be renamed from "sampling_fn" to "sample_fn"?